BERTopic inference API

davanstrien opened the pull request:
This PR adds support inference API  support for BERTopic models

davanstrien commented:
I will update `requirements.txt` once the Hugging Face integration in BERTopic has been included in a PyPI release cc @MaartenGr. I'd suggest waiting to merge this until that release is done but I think this is ready for review pending that one final change.

Narsil commented on .github/workflows/python-api-bertopic.yaml:
```diff
@@ -0,0 +1,26 @@
+on:
+  pull_request:
+    paths:
+      - "docker_images/span_marker/**"
```
```suggestion
      - "docker_images/berttopic/**"
```

Narsil commented on docker_images/bertopic/app/pipelines/__init__.py:
```diff
@@ -0,0 +1,16 @@
+from app.pipelines.tabular_classification_pipeline import TabularClassificationPipeline
+from app.pipelines.tabular_regression_pipeline import TabularRegressionPipeline
+from app.pipelines.text_to_speech import TextToSpeechPipeline
+from app.pipelines.token_classification import TokenClassificationPipeline
```
You need to remove unused imports I think

davanstrien replied:
good catch :)

osanseviero reviewed (commented):
This is looking nice!

osanseviero commented:
Btw https://github.com/huggingface/api-inference-community has some instructions to test this locally (as integration test or full end-to-end with docker), I would suggest to run the integration test locally at least :)

davanstrien commented on docker_images/bertopic/tests/test_api.py:
```diff
@@ -0,0 +1,59 @@
+# Must contain at least one example of each implemented pipeline
+# Tests do not check the actual values of the model output, so small dummy
+# models are recommended for faster tests.
+TESTABLE_MODELS: Dict[str, str] = {
```
I've updated it to parameterize inputs. I've added both cases because one model uses safetensors, and the other use Pytorch for model serialization, so I think it might be good to catch regressions in either of those scenarios.

davanstrien commented:
> Btw [huggingface/api-inference-community](https://github.com/huggingface/api-inference-community) has some instructions to test this locally (as integration test or full end-to-end with docker), I would suggest to run the integration test locally. Cool work 🔥

I have run tests locally now, and all are passing. I get some warnings from https://github.com/lmcinnes/umap/issues/1004 but those can be safely ignored.

davanstrien commented on .github/workflows/python-api-bertopic.yaml:
```diff
@@ -0,0 +1,26 @@
+          pip install --upgrade pip
+          pip install pytest pillow httpx
+          pip install -e .
+      - run: RUN_DOCKER_TESTS=1 pytest -sv tests/test_dockers.py::DockerImageTests::test_bertopic
```
Added this now

davanstrien commented:
@osanseviero hopefully addressed all comments and suggestions now. For the CI workflow, I copied an existing one and updated args with BERTopic; hopefully, that's sufficient (I'm also only testing on CPU). All other tests are passed locally. 

@MaartenGr, would you just be able to check you are happy with how I'm calling BERTopic in a way that makes sense to you [here](https://github.com/davanstrien/api-inference-community/blob/075fc61a93fa66cdfd5e4efe9bd8fbd7c574a4e8/docker_images/bertopic/app/pipelines/text_classification.py#L14)

MaartenGr commented on docker_images/bertopic/app/pipelines/text_classification.py:
```diff
@@ -0,0 +1,29 @@
+        topics, probabilities = self.model.transform(inputs)
+        results = []
+        for topic, prob in zip(topics, probabilities):
+            topic_label = self.model.generate_topic_labels()[topic]
```
If a user has manually created some custom labels, it might be worthwhile to show those instead of the default labels. Also, we can access the default topic labels with `.topic_labels_`. 

```suggestion
            if self.model.custom_labels_ is not None:
                topic_label = self.model.custom_labels_[topic + self.model_outliers]
           else:
                topic_label = self.model.topic_labels_[topic]
```

MaartenGr replied:
Other than that, looks great!

MaartenGr commented on docker_images/bertopic/app/pipelines/text_classification.py:
```diff
@@ -0,0 +1,32 @@
+        results = []
+        for topic, prob in zip(topics, probabilities):
+            if self.model.custom_labels_ is not None:
+                topic_label = self.model.custom_labels_[topic + self.model_outliers]
```
```suggestion
                topic_label = self.model.custom_labels_[topic + self.model._outliers]
```

My bad, should be a `.` there.

osanseviero commented on docker_images/bertopic/app/pipelines/text_classification.py:
```diff
@@ -0,0 +1,32 @@
+                topic_label = self.model.custom_labels_[topic + self.model._outliers]
+            else:
+                topic_label = self.model.topic_labels_[topic]
+            results.append({"label": topic_label, "score": float(prob)})
```
Do we have a test with a model with custom labels and a model with normal labels?

osanseviero reviewed (approved):
Awesome! Thanks a lot! :fire:

Narsil commented:
LGTM. Should autodeploy too. 

Shall we release ?

davanstrien commented:
> LGTM. Should autodeploy too.

thanks :) 
> Shall we release ?

I think it would be good to wait to update the pinned dependencies in `requirements.txt` -- probably can merge on Tuesday if that's okay with you.

davanstrien commented:
> > LGTM. Should autodeploy too.
> 
> thanks :)
> 
> > Shall we release ?
> 
> I think it would be good to wait to update the pinned dependencies in `requirements.txt` -- probably can merge on Tuesday if that's okay with you.

I updated pinned requirements now, so I think we should be good to merge :)
